File: src/com/aac/jiang/view/IWriter.java
package com.aac.jiang.view;

import com.aac.jiang.modle.MethodBeam;
import com.intellij.codeInsight.actions.ReformatCodeProcessor;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.project.Project;
import com.intellij.psi.*;
import com.intellij.psi.codeStyle.JavaCodeStyleManager;
import com.intellij.psi.impl.source.codeStyle.ImportHelper;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.search.PsiShortNamesCache;

import java.util.List;


/**
 * Created by YZL on 2017/8/14.
 */
public class IWriter extends WriteCommandAction.Simple {

    private MethodBeam methodBeam;
    protected PsiFile mFile;
    protected Project mProject;
    protected PsiClass mClass;
    protected PsiElementFactory mFactory;

    public IWriter(Project project, PsiFile mFile, PsiClass mClass, MethodBeam types, PsiElementFactory mFactory, PsiFile... files) {
        super(project, files);
        this.mFile = mFile;
        this.mProject = project;
        this.mClass = mClass;
        this.mFactory = mFactory;
        this.methodBeam = types;
    }

    @Override
    protected void run() {
        JavaCodeStyleManager styleManager = JavaCodeStyleManager.getInstance(mProject);
        addInitIntentMethod();
        addImport("android.content.Context", "Context");
        addImport("com.lzy.okgo.model.HttpParams", "HttpParams");
        addImport("com.aac.data.http.utils.AacUtils", "AacUtils");
        if (methodBeam.getRxType() == 0) {
            addImport("android.arch.lifecycle.LiveData", "LiveData");
            if (methodBeam.getHttpType() == 0) {
                addImport("com.aac.data.http.utils.httpLiveGet", "httpLiveGet");
            } else {
                addImport("com.aac.data.http.utils.httpLivePost", "httpLivePost");
            }
        } else {
            addImport("io.reactivex.Flowable", "Flowable");
            if (methodBeam.getHttpType() == 0) {
                addImport("com.aac.data.http.utils.httpRxGet", "httpRxGet");
            } else {
                addImport("com.aac.data.http.utils.httpRxPost", "httpRxPost");
            }
        }
        if (methodBeam.getConverterType() == 2) {
            addImport("com.aac.data.http.converter.JsonObjectConverter", "JsonObjectConverter");
            addImport("com.alibaba.fastjson.JSONObject", "JSONObject");
        } else if (methodBeam.getConverterType() == 1) {
            addImport("com.alibaba.fastjson.TypeReference", "TypeReference");
        }
        //重新格式化代码
        styleManager.optimizeImports(mFile);
        styleManager.shortenClassReferences(mClass);
        new ReformatCodeProcessor(mProject, mClass.getContainingFile(), null, false).runWithoutProgress();
    }

    private void addInitIntentMethod() {
        StringBuffer creatIntentSB = new StringBuffer();
        StringBuffer creatIntentSB2 = new StringBuffer();
        //相同部分
        if (methodBeam.getConverterType() == 0) {//bean类型
            creatIntentSB2.append(methodBeam.getBeanSName());
        } else if (methodBeam.getConverterType() == 1) {
            creatIntentSB2.append("List<");
            creatIntentSB2.append(methodBeam.getBeanSName());
            creatIntentSB2.append(">");
        } else {
            creatIntentSB2.append("JSONObject");
        }
        creatIntentSB2.append("> ");
        creatIntentSB2.append(methodBeam.getMethodName());
        creatIntentSB2.append("(Context mContext ,String string){\n");
        creatIntentSB2.append("HttpParams params =new  HttpParams();\n");
        creatIntentSB2.append("params.put(\"string\", string);");
        if (methodBeam.getConverterType() == 1) {
            creatIntentSB2.append("TypeReference typeReference = new TypeReference<List<");
            creatIntentSB2.append(methodBeam.getBeanSName());
            creatIntentSB2.append(">>() {};");
        }

        //开始方法
        creatIntentSB.append("public  ");
        if (methodBeam.getRxType() == 0) {
            creatIntentSB.append("LiveData<");
            creatIntentSB.append(creatIntentSB2.toString());
            if (methodBeam.getHttpType() == 0) {
                creatIntentSB.append("\n\t return AacUtils.httpLiveGet(");
            } else {
                creatIntentSB.append("\n\t return AacUtils.httpLivePost(");
            }
        } else {
            creatIntentSB.append("Flowable<");
            creatIntentSB.append(creatIntentSB2.toString());
            if (methodBeam.getHttpType() == 0) {
                creatIntentSB.append("\n\t return AacUtils.httpRxGet(");
            } else {
                creatIntentSB.append("\n\t return AacUtils.httpRxPost(");
            }
        }
        creatIntentSB.append(methodBeam.getUriName());
        creatIntentSB.append("\n\t\t,params,\n");
        if (methodBeam.getConverterType() == 0) {//bean类型
            creatIntentSB.append("\t\tnew BeanConverter<>(");
            creatIntentSB.append(methodBeam.getKeyName());
            creatIntentSB.append(",");
            creatIntentSB.append(methodBeam.getBeanSName());
            creatIntentSB.append(".class));\n");
        } else if (methodBeam.getConverterType() == 1) {
            creatIntentSB.append("\t\tnew BeanConverter<>(");
            creatIntentSB.append(methodBeam.getKeyName());
            creatIntentSB.append(",");
            creatIntentSB.append("typeReference.getType()));\n");
        } else {
            creatIntentSB.append("\t\tnew JsonObjectConverter());\t");
        }
        creatIntentSB.append("}\n");
        PsiMethod methodFromText = mFactory.createMethodFromText(creatIntentSB.toString(), mClass);
        mClass.addBefore(methodFromText, mClass.getLastChild());
    }


    /**
     * 添加 import
     *
     * @param fullyQualifiedName fullyQualifiedName
     * @param simpleName         simpleName
     */
    private void addImport(String fullyQualifiedName, String simpleName) {
        if (!(mFile instanceof PsiJavaFile)) {
            return;
        }
        final PsiJavaFile javaFile = (PsiJavaFile) mFile;

        PsiImportList importList = javaFile.getImportList();
        if (importList == null) {
            return;
        }

        // Check if already imported
//        for (PsiImportStatementBase is : importList.getAllImportStatements()) {
//            String impQualifiedName = is.getImportReference().getQualifiedName();
//            if (fullyQualifiedName.equals(impQualifiedName)) {
//                return; // Already imported so nothing neede
//            }
//
//        }
        if (ImportHelper.isAlreadyImported(javaFile, fullyQualifiedName)) {
            return;
        }
        GlobalSearchScope searchScope = GlobalSearchScope.allScope(mClass.getProject());
        PsiClass[] psiClasses = PsiShortNamesCache.getInstance(mClass.getProject()).getClassesByName(simpleName, searchScope);
        for (int i = 0; i < psiClasses.length; i++) {
            if (psiClasses[i].getQualifiedName().equals(fullyQualifiedName)) {
                importList.add(mFactory.createImportStatement(psiClasses[i]));
                return;
            }
        }
    }


}

File: src/com/aac/jiang/view/IWriter.kt
package cn.yzl.kotlin.ex.click

import com.aac.jiang.modle.MethodBeam
import com.intellij.codeInsight.actions.ReformatCodeProcessor
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiFile
import org.jetbrains.kotlin.idea.caches.resolve.resolveImportReference
import org.jetbrains.kotlin.idea.util.ImportInsertHelper
import org.jetbrains.kotlin.name.FqName
import org.jetbrains.kotlin.psi.KtClass
import org.jetbrains.kotlin.psi.KtFile
import org.jetbrains.kotlin.psi.KtPsiFactory

/**
 * Created by YZL on 2017/8/14.
 */
class IWriter(protected var mProject: Project, protected var mFile: PsiFile, protected var mClass: KtClass,
              private val methodBeam: MethodBeam)
    : WriteCommandAction.Simple<Any>(mProject, mFile) {
    var ktPsiFactory: KtPsiFactory

    init {
        ktPsiFactory = KtPsiFactory(mProject, false)
    }

    @Throws(Throwable::class)
    override fun run() {
        createCode()
//       重新格式化代码
        ReformatCodeProcessor(mProject, mFile, null, true).run()
    }

    private fun createCode() {
        run {
            //onclick(v:View)
            val creatIntentSB = StringBuffer()
            val creatIntentSB2 = StringBuffer()
            when {
                methodBeam.converterType == 0 -> //bean类型
                    creatIntentSB2.append(methodBeam.beanSName)
                methodBeam.converterType == 1 -> {
                    creatIntentSB2.append("List<")
                    creatIntentSB2.append(methodBeam.beanSName)
                    creatIntentSB2.append(">")
                }
                else -> creatIntentSB2.append("JSONObject")
            }
            creatIntentSB2.append(">{\n")
            creatIntentSB2.append("\tval params = HttpParams()\n")
            creatIntentSB2.append("\tparams.put(\"string\", string)")
            if (methodBeam.converterType == 1) {
                creatIntentSB2.append("\n\tval typeReference =  object : TypeReference<List<")
                creatIntentSB2.append(methodBeam.beanSName)
                creatIntentSB2.append(">>() {}")
            }

            //方法开始
            creatIntentSB.append("fun ")
            creatIntentSB.append(methodBeam.methodName)
            creatIntentSB.append("(context: Context,string: String):")
            if (methodBeam.rxType == 0) {
                creatIntentSB.append("LiveData<")
                creatIntentSB.append(creatIntentSB2.toString())
                if (methodBeam.httpType == 0) {
                    creatIntentSB.append("\n\t return httpLiveGet<")
                } else {
                    creatIntentSB.append("\n\t return httpLivePost<")
                }
            } else {
                creatIntentSB.append("Flowable<")
                creatIntentSB.append(creatIntentSB2.toString())
                if (methodBeam.httpType == 0) {
                    creatIntentSB.append("\n\t return httpRxGet<")
                } else {
                    creatIntentSB.append("\n\t return httpRxPost<")
                }
            }
            creatIntentSB.append(methodBeam.beanSName)
            creatIntentSB.append(">(")
            creatIntentSB.append(methodBeam.uriName)
            creatIntentSB.append("\n\t\t\t,params,\n")
            when {
                methodBeam.converterType == 0 -> {//bean类型
                    creatIntentSB.append("\t\t\tBeanConverter(")
                    creatIntentSB.append(methodBeam.keyName)
                    creatIntentSB.append(",")
                    creatIntentSB.append(methodBeam.beanSName)
                    creatIntentSB.append("::class.java))\n")
                }
                methodBeam.converterType == 1 -> {
                    creatIntentSB.append("\t\t\tBeanConverter(")
                    creatIntentSB.append(methodBeam.keyName)
                    creatIntentSB.append(",")
                    creatIntentSB.append("typeReference.type))\n")
                }
                else -> creatIntentSB.append("\t\t\tJsonObjectConverter())\t")
            }
            creatIntentSB.append("\n}")
            if (mClass.getBody() != null && mClass.getBody()!!.lastChild != null) {
                try {
                    mClass.getBody()!!.addBefore(ktPsiFactory.createFunction(creatIntentSB.toString()),
                            mClass.getBody()!!.lastChild)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
        addImports()
    }

    //添加引用路径
    fun addImports() {
        insertImports(mClass.containingKtFile, "android.content.Context")
        insertImports(mClass.containingKtFile, "com.lzy.okgo.model.HttpParams")
        if (methodBeam.rxType == 0) {
            insertImports(mClass.containingKtFile, "android.arch.lifecycle.LiveData")
            if (methodBeam.httpType == 0) {
                insertImports(mClass.containingKtFile, "com.aac.data.http.utils.httpLiveGet")
            } else {
                insertImports(mClass.containingKtFile, "com.aac.data.http.utils.httpLivePost")
            }
        } else {
            insertImports(mClass.containingKtFile, "io.reactivex.Flowable")
            if (methodBeam.httpType == 0) {
                insertImports(mClass.containingKtFile, "com.aac.data.http.utils.httpRxGet")
            } else {
                insertImports(mClass.containingKtFile, "com.aac.data.http.utils.httpRxPost")
            }
        }
        if (methodBeam.converterType == 2) {
            insertImports(mClass.containingKtFile, "com.aac.data.http.converter.JsonObjectConverter")
            insertImports(mClass.containingKtFile, "com.alibaba.fastjson.JSONObject")
        } else if (methodBeam.converterType == 1) {
            insertImports(mClass.containingKtFile, "com.alibaba.fastjson.TypeReference")
        }
    }

    fun insertImports(ktFile: KtFile, path: String) {
        ktFile.importList
        val importList = ktFile.importDirectives
        for (importDirective in importList) {
            val importPath = importDirective.importPath
            if (importPath != null) {
                val pathStr = importPath.pathStr
                if (pathStr == path) {
                    return
                }
            }
        }
        ImportInsertHelper.getInstance(mProject)
                .importDescriptor(ktFile, ktFile.resolveImportReference(FqName(path)).iterator().next(), false)
    }


}
